lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskHostBase.cs

```diff
@@ -0,0 +1,242 @@
+// under the License.
+
+using System;
+using System.IO;
+using System.Net.Sockets;
+using System.Runtime.Remoting;
+using System.Threading;
+using Org.Apache.REEF.Common.Tasks.Events;
+using Org.Apache.REEF.IMRU.OnREEF.Driver;
+using Org.Apache.REEF.Network.Group.Task;
+using Org.Apache.REEF.Utilities.Logging;
+using Org.Apache.REEF.Wake.Remote.Impl;
+
+namespace Org.Apache.REEF.IMRU.OnREEF.IMRUTasks
+{
+    internal abstract class TaskHostBase
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(TaskHostBase));
+
+        /// <summary>
+        /// Specify whether to invoke garbage collector or not
+        /// </summary>
+        protected readonly bool InvokeGc;
+
+        /// <summary>
+        /// Shows if the object has been disposed.
+        /// </summary>
+        private int _disposed = 0;
+
+        /// <summary>
+        /// Group Communication client for the task
+        /// </summary>
+        private readonly IGroupCommClient _groupCommunicationsClient;
+
+        /// <summary>
+        /// CommunicationGroupClient for the task
+        /// </summary>
+        protected readonly ICommunicationGroupClient _communicationGroupClient;
+
+        /// <summary>
+        /// Task close Coordinator to handle the work when receiving task close event
+        /// </summary>
+        private readonly TaskCloseCoordinator _taskCloseCoordinator;
+
+        /// <summary>
+        /// The cancellation token to control the group communication operation cancellation
+        /// </summary>
+        protected readonly CancellationTokenSource CancellationSource;
+
+        /// <summary>
+        /// Task host base class to hold the common stuff of both map and update tasks
+        /// </summary>
+        /// <param name="groupCommunicationsClient"></param>
+        /// <param name="taskCloseCoordinator"></param>
+        /// <param name="invokeGc"></param>
+        protected TaskHostBase(
+            IGroupCommClient groupCommunicationsClient,
+            TaskCloseCoordinator taskCloseCoordinator,
+            bool invokeGc)
+        {
+            _groupCommunicationsClient = groupCommunicationsClient;
+            _communicationGroupClient = groupCommunicationsClient.GetCommunicationGroup(IMRUConstants.CommunicationGroupName);
+
+            InvokeGc = invokeGc;
+            _taskCloseCoordinator = taskCloseCoordinator;
+            CancellationSource = new CancellationTokenSource();
+        }
+
+        /// <summary>
+        /// Handle the exceptions in the Call() method
+        /// Default to IMRUSystemException to make it recoverable
+        /// </summary>
+        /// <param name="memento"></param>
+        /// <returns></returns>
+        public byte[] Call(byte[] memento)
+        {
+            Logger.Log(Level.Info, "Entering {0} Call().", TaskHostName);
+            try
+            {
+                return TaskBody(memento);
+            }
+            catch (OperationCanceledException e)
```

Comment:
Please gather all clauses which do `HanldeCommunicationException` in one, as described at http://stackoverflow.com/questions/136035/catch-multiple-exceptions-at-once. Personally I like `catch (Exception ex) when (ex is ...)` version better